libnest/src/error.rs

```diff
@@ -0,0 +1,206 @@
+pub enum CacheErrorKind {
+    /// The error occured doing an IO operation
+    #[fail(display = "{}", _0)]
+    IO(String),
+    /// The error occured when deserializing a manifest
+    #[fail(display = "{}", _0)]
+    Deserialize(String),
+    /// The error occured when serializing a manifest
+    #[fail(display = "{}", _0)]
+    Serialize(String),
+}
+
+/// A type for errors that may occure when using the manifests cache.
+#[derive(Debug)]
+pub struct CacheError {
+    inner: Context<CacheErrorKind>,
+}
+
+impl CacheError {
+    /// Returns a [`CacheErrorKind`] the reason why this error was thrown
+    pub fn kind(&self) -> &CacheErrorKind {
+        &self.inner.get_context()
+    }
+}
+
+impl Fail for CacheError {
+    fn cause(&self) -> Option<&Fail> {
+        self.inner.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.inner.backtrace()
+    }
+}
+
+impl Display for CacheError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Display::fmt(&self.inner, f)
+    }
+}
+
+impl From<CacheErrorKind> for CacheError {
+    fn from(kind: CacheErrorKind) -> CacheError {
+        CacheError {
+            inner: Context::new(kind),
+        }
+    }
+}
+
+impl From<Context<CacheErrorKind>> for CacheError {
+    fn from(inner: Context<CacheErrorKind>) -> CacheError {
+        CacheError { inner }
+    }
+}
+
+/// Kind of errors that may occure when pulling a repository
+#[derive(Clone, PartialEq, Debug, Fail)]
+pub enum PullErrorKind {
+    /// The error is network-based and occured while downloading the datas
+    #[fail(display = "{}", _0)]
+    Download(String),
+    /// The error occured while interpreting the received data
+    #[fail(display = "{}: received invalid data", _0)]
+    InvalidData(Url),
+    /// The error occured while trying to remove the old cache
+    #[fail(display = "can't remove old cache: {}", _0)]
+    CantRemoveCache(String),
+    /// The error occured while trying to create the new cache
+    #[fail(display = "can't create cache: {}", _0)]
+    CantCreateCache(String),
+}
+
+/// A type for errors that may occure when pulling a repository
+#[derive(Debug)]
+pub struct PullError {
+    inner: Context<PullErrorKind>,
+}
+
+impl Fail for PullError {
+    fn cause(&self) -> Option<&Fail> {
+        self.inner.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.inner.backtrace()
+    }
+}
+
+impl Display for PullError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Display::fmt(&self.inner, f)
+    }
+}
+
+impl From<PullErrorKind> for PullError {
+    fn from(kind: PullErrorKind) -> PullError {
+        PullError {
+            inner: Context::new(kind),
+        }
+    }
+}
+
+impl From<Context<PullErrorKind>> for PullError {
+    fn from(inner: Context<PullErrorKind>) -> PullError {
+        PullError { inner }
+    }
+}
+
+/// Kind of errors that may occure when downloading a package from a repository
```

Comment:
This sentence does not it your usual style, but there may be a reason I don't see for that.